feat: ic binary dependencies are included in tarball

smallstepman opened the pull request:
continuation of #39 which got reverted in #43

ericswanson-dfinity commented on extensions-utils/src/dfx.rs:
```diff
@@ -50,6 +54,57 @@ where
+    // The same applies to other bundled binaries.
+    command.env(
+        "PATH",
+        binary_to_call.parent().unwrap_or_else(|| Path::new(".")),
```
Isn't `binary_to_call.parent()` by definition `extension_dir_path` ?
```suggestion
        extension_dir_path.unwrap_or_else(|| Path::new(".")),
```

ericswanson-dfinity commented on extensions-utils/src/dfx.rs:
```diff
@@ -50,6 +54,57 @@ where
+{
+    let extension_binary_path =
+        std::env::current_exe().map_err(|e| anyhow::anyhow!("Failed to get current exe: {}", e))?;
+    let extension_dir_path = extension_binary_path
+        .parent()
+        .unwrap_or_else(|| Path::new("."));
```
Can we treat this as an error?  Is there a circumstance in which we would want to actually continue here?

ericswanson-dfinity commented on extensions-utils/src/dfx.rs:
```diff
@@ -50,6 +54,57 @@ where
+            if output.status.success() {
+                Ok(String::from_utf8_lossy(&output.stdout).into_owned())
+            } else {
+                let args: Vec<_> = command
```
Please take a look at the '"npm is not installed"'` test in https://github.com/dfinity/sdk/blob/master/e2e/tests-dfx/error_context.bash#L127

Change `assert_contains 'program: "npm"'` to check for something else, or otherwise take a look at the error output.

The rust 1.71.1 update brought along pretty detailed `Command` output.  It would be worthwhile to see if we can make use of that here, without rolling our own.

smallstepman replied:
this one is still TODO

ericswanson-dfinity commented on extensions-utils/src/download_dependencies.rs:
```diff
@@ -0,0 +1,44 @@
+    );
+    println!("Downloading {}", url);
+
+    let resp = reqwest::blocking::get(&url).expect("Request failed");
```
Please add retry logic

ericswanson-dfinity replied:
Recommendation: use the backoff crate, for example https://github.com/dfinity/sdk/pull/3255/files

ericswanson-dfinity commented on extensions-utils/src/download_dependencies.rs:
```diff
@@ -0,0 +1,44 @@
+    #[cfg(unix)]
+    {
+        use std::os::unix::fs::PermissionsExt;
+        dfx_core::fs::set_permissions(&out_path, std::fs::Permissions::from_mode(0o777))
```
`0o500` would be better  (owner read/execute)

ericswanson-dfinity commented on extensions-utils/src/dfx.rs:
```diff
@@ -50,6 +54,57 @@ where
+    // The sns command line tool itself calls dfx; it should call this dfx.
+    // The sns command line tool should not rely on commands not packaged with dfx.
+    // The same applies to other bundled binaries.
+    command.env(
```
where is the path to dfx coming from?  Some of the use cases for dfx extensions have the extension call dfx one or more times.  It used to be the case, and may still be, that `sns-cli`, `ic-admin`, and/or `ic-nns-init` called dfx.  For this reason we talked about passing the path to dfx to the extension when calling it.   (It's understood that the experience might not be great if this means asking for an identity passcode more than once).

smallstepman replied:
right, forgot about it

ericswanson-dfinity commented on extensions/nns/src/install_nns.rs:
```diff
@@ -416,38 +412,24 @@ pub struct IcNnsInitOpts {
+        "--url",
+        &opts.nns_url,
+        "--wasm-dir",
+        opts.wasm_dir.to_str().unwrap(),
```
Can you think of a way to do this without a lossy conversion from PathBuf to String?

smallstepman replied:
turned everything into OsString

ericswanson-dfinity commented on extensions/nns/src/install_nns.rs:
```diff
@@ -416,38 +412,24 @@ pub struct IcNnsInitOpts {
+        args.push("--sns-subnet");
+        args.push(subnets);
+    }
+    call_extension_bundled_binary("ic-nns-init", &args)
+        .with_context(|| format!("Error executing `ic-admin` with args: {:?}", args))
```
calling ic-nns-init but reporting an error with ic-admin

ericswanson-dfinity commented on extensions/nns/src/install_nns.rs:
```diff
@@ -520,37 +488,24 @@ pub fn upload_nns_sns_wasms_canister_wasms(dfx_cache_path: &Path) -> anyhow::Res
-            })?;
+        let wasm_path = nns_wasm_dir(dfx_cache_path)
+            .join(wasm_name)
+            .display()
+            .to_string();
```
Please find a way to do this without a lossy conversion from PathBuf.  `PathBuf.display()` is for display, not for passing to arguments or ultimately using to open a file.

smallstepman replied:
turned everything into OsString

ericswanson-dfinity reviewed (changes requested):
Several comments that you responded to here: https://github.com/dfinity/dfx-extensions/pull/39 are not addressed in this PR.  I re-added some, but may have missed others.

smallstepman commented:
> Several comments that you responded to here: https://github.com/dfinity/dfx-extensions/pull/39 are not addressed in this PR. I re-added some, but may have missed others.

Oops, I should've marked PR as draft. At this point, this is just a revert of a revert. I'll soon push a commit that addresses your review from other PR

smallstepman commented on extensions/nns/Cargo.toml:
```diff
@@ -46,5 +50,4 @@ pre-release-replacements = [
 [package.metadata.dist]
-include = ["extension.json"]
-
+include = ["extension.json", "sns-cli", "ic-admin", "ic-nns-init"]
```
I just played with the idea, but it would be a lot of work (we would need to repackage everything during e2e tests and release CI manually). I've been closely monitoring the cargo-dist project for a while now; as soon as they'll make a release supporting this functionality, I'll bring it here.

ericswanson-dfinity replied:
Sounds good!

ericswanson-dfinity commented on extensions-utils/src/dependencies/call.rs:
```diff
@@ -0,0 +1,43 @@
+    })?;
+    let binary_to_call = extension_dir_path.join(binary_name);
+    let mut command = Command::new(&binary_to_call);
+    // If extension's dependency calls dfx; it should call dfx in this dir.
+    command.env("PATH", dfx_cache_path.join("dfx"));
```
I still don't understand removing the rest of the path.  This would make `dfx build/deploy` fail when trying to call `npm`, for one. `dfx extension run` doesn't do this, though it does pass `--dfx-cache-path`.

Could we leave the PATH alone here, and let the extension pass through `--dfx-cache-path` if it needs to?

smallstepman replied:
From rust docs https://doc.rust-lang.org/std/process/struct.Command.html#method.env
> pub fn [env](https://doc.rust-lang.org/std/process/struct.Command.html#method.env)<K, V>(&mut self, key: K, val: V) -> &mut [Command](https://doc.rust-lang.org/std/process/struct.Command.html)
...
Inserts or updates an explicit environment variable mapping.

but looks like I misinterpreted `updates` for `appends`, just tested it and indeed it overwrites the content of PATH which is not what I thought it will do

smallstepman replied:
From rust docs https://doc.rust-lang.org/std/process/struct.Command.html#method.env
> pub fn [env](https://doc.rust-lang.org/std/process/struct.Command.html#method.env)<K, V>(&mut self, key: K, val: V) -> &mut [Command](https://doc.rust-lang.org/std/process/struct.Command.html)
...
Inserts or updates an explicit environment variable mapping.

but looks like I misinterpreted `updates` for `appends`, just tested it and indeed it overwrites the content of PATH which is not what I thought it will do

smallstepman replied:
> Could we leave the PATH alone here, and let the extension pass through --dfx-cache-path if it needs to?

I think this is not entirely up to us, as it would require upstream changes in command-line interfaces of ic-admin ic-nns-init and sns-cli. The solution you're proposing is certainly more robust, but let's come back to this in different PR https://dfinity.atlassian.net/browse/SDK-1201

ericswanson-dfinity commented on extensions-utils/src/dependencies/call.rs:
```diff
@@ -0,0 +1,43 @@
+/// Calls a binary that was delivered with an extension tarball.
+///
+/// # Returns
+/// - On success, returns stdout as a string.
```
Is stdout also passed through (displayed) as the binary runs?  If so or if not, can we say here what the behavior is?

ericswanson-dfinity commented on extensions-utils/src/dependencies/download_ic_binaries.rs:
```diff
@@ -0,0 +1,64 @@
+    let mut temp = fs::File::create(&temp_file).expect("Failed to create the file");
+    copy(&mut d, &mut temp).expect("Failed to copy content");
+
+    fs::rename(temp_file, &destination_path).expect("Failed to move extension");
+    #[cfg(unix)]
+    {
+        use std::os::unix::fs::PermissionsExt;
+        dfx_core::fs::set_permissions(&destination_path, std::fs::Permissions::from_mode(0o500))
+            .expect("Failed to set permissions");
+    }
```
If we set the permission first, then we either get an executable downloaded binary in destination path, or nothing, but we never get a not-executable downloaded binary in the destination path.
```suggestion
    #[cfg(unix)]
    {
        use std::os::unix::fs::PermissionsExt;
        dfx_core::fs::set_permissions(&destination_path, std::fs::Permissions::from_mode(0o500))
            .expect("Failed to set permissions");
    }
    fs::rename(temp_file, &destination_path).expect("Failed to move extension");
```

ericswanson-dfinity commented on extensions/nns/build.rs:
```diff
@@ -0,0 +1,36 @@
+        .join("target")
+        .join(std::env::var("PROFILE").unwrap());
+    for (binary_name, renamed_binary_name) in BINARY_DEPENDENCIES {
+        let destination_paths = (
+            manifest_dir.join(renamed_binary_name),
+            target_dir.join(renamed_binary_name),
+        );
```
We can improve readability by giving meaningful names to `destination_paths.0` and `destination_paths.1`.  This construct as-written is essentially `let local_variables = (first local variable, second local variable)` and then accessing `local_variables.0` and `local_variables.1`.

Also, not to put too fine a point on it, but the final act of this method is to copy one "destination path" to the other "destination path".

I leave the details to the PR author.

```suggestion
        let some_meaningful_name = manifest_dir.join(renamed_binary_name);
        let some_other_meaningful_name = target_dir.join(renamed_binary_name);
```

ericswanson-dfinity commented on extensions/nns/build.rs:
```diff
@@ -0,0 +1,36 @@
+        if destination_paths.1.exists() {
+            std::fs::remove_file(&destination_paths.1).unwrap();
+        }
+        std::fs::create_dir_all(destination_paths.1.parent().unwrap()).unwrap();
```
`destination_paths.1.parent().unwrap()` is by definition `target_dir`
```suggestion
        std::fs::create_dir_all(target_dir).unwrap();
```

ericswanson-dfinity commented on extensions/nns/src/install_nns.rs:
```diff
@@ -416,38 +413,24 @@ pub struct IcNnsInitOpts {
+        args.push(subnets.into());
+    }
+    call_extension_bundled_binary(dfx_cache_path, "ic-nns-init", &args)
+        .with_context(|| format!("Error executing `ic-nns-init` with args: {:?}", args))
```
The error output for `call_extension_bundled_binary` already includes the arguments, yes?

ericswanson-dfinity commented on extensions/nns/src/install_nns.rs:
```diff
@@ -520,37 +493,21 @@ pub fn upload_nns_sns_wasms_canister_wasms(dfx_cache_path: &Path) -> anyhow::Res
+            upload_name.into(),
+        ];
+        call_extension_bundled_binary(dfx_cache_path,"sns-cli", &args)
+            .map_err(|e| anyhow!(
```
call_extension_bundled_binary error should already include the args

ericswanson-dfinity commented on extensions/sns/build.rs:
```diff
@@ -0,0 +1,34 @@
+        .join("target")
+        .join(std::env::var("PROFILE").unwrap());
+    for (binary_name, renamed_binary_name) in BINARY_DEPENDENCIES {
+        let destination_paths = (
+            manifest_dir.join(renamed_binary_name),
+            target_dir.join(renamed_binary_name),
+        );
```
same issue as other build.rs

ericswanson-dfinity commented on extensions/sns/build.rs:
```diff
@@ -0,0 +1,34 @@
+    // keep copy of the dependency in the root of the project, so that cargo-dist will be able to package it into a tarball
+    let manifest_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
+    // and also in `target/debug` or `target/release` for development purposes (e.g. cargo run)
+    let target_dir = manifest_dir
```
I'm no expert, but is this `CARGO_TARGET_DIR`?

smallstepman replied:
it is `CARGO_TARGET_DIR`, but `CARGO_TARGET_DIR` is not available when running build script

ericswanson-dfinity commented on extensions/sns/build.rs:
```diff
@@ -0,0 +1,34 @@
+        .parent()
+        .unwrap()
+        .join("target")
+        .join(std::env::var("PROFILE").unwrap());
```
> PROFILE — release for release builds, debug for other builds. This is determined based on if the [profile](https://doc.rust-lang.org/cargo/reference/profiles.html) inherits from the [dev](https://doc.rust-lang.org/cargo/reference/profiles.html#dev) or [release](https://doc.rust-lang.org/cargo/reference/profiles.html#release) profile. Using this environment variable is not recommended. Using other environment variables like OPT_LEVEL provide a more correct view of the actual settings being used.

> Using this environment variable is not recommended. 

Is there a `CARGO_*` environment variable that is the whole directory we need here?

smallstepman replied:
Unfortunately, no: https://github.com/rust-lang/cargo/issues/9661
I printed every envvar available during build script with
```rust
    for (key, value) in std::env::vars() {
        println!("{}: {}", key, value);
    }
```
and there is not much to choose from :(. The only other good option is taking a parent.parent.parent of `OUT_DIR`:
```
  CARGO: /Users/mnl/.rustup/toolchains/stable-aarch64-apple-darwin/bin/cargo
  CARGO_CFG_PANIC: unwind
  CARGO_CFG_TARGET_ARCH: aarch64
  CARGO_CFG_TARGET_ENDIAN: little
  CARGO_CFG_TARGET_ENV:
  CARGO_CFG_TARGET_FAMILY: unix
  CARGO_CFG_TARGET_FEATURE: aes,crc,dit,dotprod,dpb,dpb2,fcma,fhm,flagm,fp16,frintts,jsconv,lor,lse,neon,paca,pacg,pan,pmuv3,ras,rcpc,rcpc2,rdm,sb,sha2,sha3,ssbs,vh
  CARGO_CFG_TARGET_HAS_ATOMIC: 128,16,32,64,8,ptr
  CARGO_CFG_TARGET_OS: macos
  CARGO_CFG_TARGET_POINTER_WIDTH: 64
  CARGO_CFG_TARGET_VENDOR: apple
  CARGO_CFG_UNIX:
  CARGO_ENCODED_RUSTFLAGS:
  CARGO_HOME: /Users/mnl/.cargo
  CARGO_MAKEFLAGS: -j --jobserver-fds=12,13 --jobserver-auth=12,13
  CARGO_MANIFEST_DIR: /Users/mnl/Desktop/dfx-extensions/extensions/sns
  CARGO_PKG_AUTHORS: DFINITY Team
  CARGO_PKG_DESCRIPTION:
  CARGO_PKG_HOMEPAGE:
  CARGO_PKG_LICENSE: Apache-2.0
  CARGO_PKG_LICENSE_FILE:
  CARGO_PKG_NAME: sns
  CARGO_PKG_README: README.md
  CARGO_PKG_REPOSITORY: https://github.com/dfinity/dfx-extensions
  CARGO_PKG_RUST_VERSION:
  CARGO_PKG_VERSION: 0.1.0
  CARGO_PKG_VERSION_MAJOR: 0
  CARGO_PKG_VERSION_MINOR: 1
  CARGO_PKG_VERSION_PATCH: 0
  CARGO_PKG_VERSION_PRE:
  DEBUG: true
  DYLD_FALLBACK_LIBRARY_PATH: /Users/mnl/Desktop/dfx-extensions/target/debug/deps:/Users/mnl/Desktop/dfx-extensions/target/debug:/Users/mnl/.rustup/toolchains/stable-aarch64-apple-darwin/lib/rustlib/aarch64-apple-darwin/lib:/Users/mnl/.rustup/toolchains/stable-aarch64-apple-darwin/lib:/Users/mnl/lib:/usr/local/lib:/usr/lib:/Users/mnl/Desktop/dfx-extensions/target/debug/build/libloading-482d06ac5b900652/out:/Users/mnl/Desktop/dfx-extensions/target/debug/build/libz-ng-sys-0f0fa38f8a3c41b2/out/lib:/Users/mnl/Desktop/dfx-extensions/target/debug/build/libz-ng-sys-0f0fa38f8a3c41b2/out/lib64:/Users/mnl/Desktop/dfx-extensions/target/debug/build/psm-9ab121517395d35c/out:/Users/mnl/Desktop/dfx-extensions/target/debug/build/ring-483e819f595b7e9c/out
  HOME: /Users/mnl
  HOST: aarch64-apple-darwin
  OLDPWD: /Users/mnl
  OPT_LEVEL: 0
  OUT_DIR: /Users/mnl/Desktop/dfx-extensions/target/debug/build/sns-5365b70adbf9ff15/out
  PROFILE: debug
  PWD: /Users/mnl/Desktop/dfx-extensions
  RUSTC: /Users/mnl/.rustup/toolchains/stable-aarch64-apple-darwin/bin/rustc
  RUSTDOC: /Users/mnl/.rustup/toolchains/stable-aarch64-apple-darwin/bin/rustdoc
  RUSTUP_HOME: /Users/mnl/.rustup
  RUSTUP_TOOLCHAIN: stable-aarch64-apple-darwin
  RUST_RECURSION_COUNT: 1
  TARGET: aarch64-apple-darwin
```
(I removed system-related envvars)

smallstepman replied:
I went with `OUT_DIR`, since `PROFILE` is not recommended (which I didn't know, thanks)

smallstepman commented on extensions-utils/src/dependencies/dfx.rs:
```diff
@@ -1,53 +1,38 @@
     let binary = dfx_cache_path.join(command);
-
     let mut command = Command::new(&binary);
+    // If extension's dependency calls dfx; it should call dfx in this dir.
+    command.env("PATH", dfx_cache_path.join("dfx"));
```
I went with prepending for now. We can address this issue in a separate PR

ericswanson-dfinity commented on extensions/sns/build.rs:
```diff
@@ -0,0 +1,34 @@
+        if bin_in_target_dir.exists() {
+            std::fs::remove_file(&bin_in_target_dir).unwrap();
+        }
+        std::fs::create_dir_all(bin_in_target_dir.parent().unwrap()).unwrap();
```
```suggestion
        std::fs::create_dir_all(target_dir).unwrap();
```

ericswanson-dfinity commented on extensions/nns/build.rs:
```diff
@@ -0,0 +1,36 @@
+        if bin_in_target_dir.exists() {
+            std::fs::remove_file(&bin_in_target_dir).unwrap();
+        }
+        std::fs::create_dir_all(bin_in_target_dir.parent().unwrap()).unwrap();
```
```suggestion
        std::fs::create_dir_all(target_dir).unwrap();
```

smallstepman replied:
missed that, thanks

smallstepman commented on extensions-utils/src/dependencies/call.rs:
```diff
@@ -0,0 +1,53 @@
+/// - On success, returns stdout as a string.
+/// - On error, returns an error message including stdout and stderr.
+///
+/// Does not stream stdout/stderr, and instead returns it after the process has exited.
```
```suggestion
/// Does not print stdout/stderr to the console, and instead returns the output to the caller after the process has exited.
```

smallstepman commented on extensions-utils/src/dependencies/dfx.rs:
```diff
@@ -1,53 +1,48 @@
 /// - On success, returns stdout as a string.
 /// - On error, returns an error message including stdout and stderr.
+///
+/// Does not stream stdout/stderr, and instead returns it after the process has exited.
```
```suggestion
/// Does not print stdout/stderr to the console, and instead returns the output to the caller after the process has exited.
```

ericswanson-dfinity commented on extensions-utils/src/dependencies/call.rs:
```diff
@@ -0,0 +1,53 @@
+        )
+    })?;
+    let binary_to_call = extension_dir_path.join(binary_name);
+    let mut command = Command::new(binary_to_call);
+    // If extension's dependency calls dfx; it should call dfx in this dir.
+    if let Some(path) = env::var_os("PATH") {
+        let mut paths = env::split_paths(&path).collect::<Vec<_>>();
+        paths.push(dfx_cache_path.to_path_buf());
+        let new_path = env::join_paths(paths)?;
+        command.env("PATH", new_path);
+    } else {
+        command.env("PATH", dfx_cache_path);
+    }
+    command.args(args);
+    let output = command
+        .stdin(process::Stdio::null())
+        .output()
+        .with_context(|| format!("Error executing {:#?}", command))?
+        .stdout;
+    Ok(String::from_utf8_lossy(&output).to_string())
```
Is there an opportunity to apply DRY between this method and `call_dfx_bundled_binary`?
